File: packages/python/src/output_guard/formats/detect.py
"""detect-format: heuristic format detection.

Mirror of packages/ts/src/formats/detect.ts.
"""

from __future__ import annotations

import re
from typing import Literal

_TOML_HEADER = re.compile(r"^\[[\w.]+\]")
_TOML_KV = re.compile(r"^[\w-]+\s*=\s*")
_YAML_KV = re.compile(r"^[\w-]+\s*:\s*")
_PYTHON_TOKEN = re.compile(r"\bTrue\b|\bFalse\b|\bNone\b")

ConcreteFormat = Literal["json", "yaml", "toml", "python"]


def detect_format(text: str) -> ConcreteFormat:
    trimmed = text.lstrip()
    if _TOML_HEADER.match(trimmed) or _TOML_KV.match(trimmed):
        return "toml"
    if trimmed.startswith("---") or (
        _YAML_KV.match(trimmed) and not trimmed.startswith("{")
    ):
        return "yaml"
    if trimmed.startswith("{") or trimmed.startswith("["):
        return "json"
    if trimmed.startswith("(") or _PYTHON_TOKEN.search(trimmed):
        return "python"
    return "json"

File: packages/ts/src/formats/detect.ts
/**
 * detect format: heuristic detection of input format.
 * Priority: TOML (key=val pattern), YAML (leading ---), JSON ({ or [), python-literal.
 */
import type { Format } from "../results.js";

export const detectFormat = (text: string): Exclude<Format, "auto"> => {
  const trimmed = text.trimStart();

  // TOML: starts with [section] header or key=val
  if (/^\[[\w.]+\]/.test(trimmed) || /^[\w-]+\s*=\s*/.test(trimmed)) {
    return "toml";
  }

  // YAML: starts with --- or key: value (but not { or [)
  if (
    trimmed.startsWith("---") ||
    (/^[\w-]+\s*:\s*/.test(trimmed) && !trimmed.startsWith("{"))
  ) {
    return "yaml";
  }

  // JSON: starts with { or [
  if (trimmed.startsWith("{") || trimmed.startsWith("[")) {
    return "json";
  }

  // Python literal: starts with { or [ after pythonic keywords
  if (trimmed.startsWith("(") || /\bTrue\b|\bFalse\b|\bNone\b/.test(trimmed)) {
    return "python";
  }

  return "json";
};
